File: kotlin/scrabble-score/src/main/kotlin/ScrabbleScore.kt
object ScrabbleScore {


    private val MAP = mapOf('A' to 1, 'B' to 3, 'C' to 3, 'D' to 2, 'E' to 1, 'F' to 4,
            'G' to 2, 'H' to 4, 'I' to 1, 'J' to 8, 'K' to 5, 'L' to 1,
            'M' to 3, 'N' to 1, 'O' to 1, 'P' to 3, 'Q' to 10, 'R' to 1,
            'S' to 1, 'T' to 1, 'U' to 1, 'V' to 4, 'W' to 4, 'X' to 8,
            'Y' to 4, 'Z' to 10)

    private fun scoreLetter(c: Char): Int {

        return MAP[c]!!
    }

    fun scoreWord(word: String): Int {
        return word.toUpperCase().toCharArray().map { scoreLetter(it) }.sum()
    }
}

File: fsharp/scrabble-score/ScrabbleScore.fs
module ScrabbleScore

open System

let private letterToScore =
    function
    | 'D' | 'G' -> 2
    | 'B' | 'C' | 'M' | 'P' -> 3
    | 'F' | 'H' | 'V' | 'W' | 'Y' -> 4
    | 'K' -> 5
    | 'J' | 'X' -> 8
    | 'Q' | 'Z' -> 10
    | _ -> 1

let score word = word |> Seq.map (fun c -> c |> Char.ToUpper |> letterToScore) |> Seq.sum
